Deployment: Fully Respect ControllerRef

enisoc opened the pull request:
**What this PR does / why we need it**:

This is part of the completion of the [ControllerRef](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/controller-ref.md) proposal. It brings Deployment into full compliance with ControllerRef. See the individual commit messages for details.

**Which issue this PR fixes**:

This ensures that Deployment does not fight with other controllers over control of Pods and ReplicaSets.

Ref: https://github.com/kubernetes/kubernetes/issues/24433

**Special notes for your reviewer**:

**Release note**:

```release-note
Deployment now fully respects ControllerRef to avoid fighting over Pods and ReplicaSets. At the time of upgrade, **you must not have Deployments with selectors that overlap**, or else [ownership of ReplicaSets may change](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/controller-ref.md#upgrading).
```
cc @erictune @kubernetes/sig-apps-pr-reviews

Commit:
Deployment: Use ControllerRef to list controlled objects.

Although Deployment already applied its ControllerRef to adopt matching
ReplicaSets, it actually still used label selectors to list objects that
it controls. That meant it didn't actually follow the rules of
ControllerRef, so it could still fight with other controller types.

This should mean that the special handling for overlapping Deployments
is no longer necessary, since each Deployment will only see objects that
it owns (via ControllerRef).

Commit:
Deployment/util: Filter by ControllerRef.

The list functions in deployment/util are used outside the Deployment
controller itself. Therefore, they don't do actual adoption/orphaning.
However, they still need to avoid listing things that don't belong.

Commit:
Deployment: Use ControllerRef to route watch events.

This is part of the completion of ControllerRef, as described here:

https://github.com/kubernetes/community/blob/master/contributors/design-proposals/controller-ref.md#watches

Commit:
Deployment: Update Lister documentation for ControllerRef.

The Deployment Listers still use selectors, because this is the
behavior expected by callers. This clarifies the meaning of the
returned list. Some callers may need to switch to using
GetControllerOf() instead, but that is a separate, case-by-case issue.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/42175)
<!-- Reviewable:end -->

kargakis commented on pkg/controller/deployment/deployment_controller.go:
```diff
@@ -58,9 +55,8 @@ const (
-	return extensions.SchemeGroupVersion.WithKind("Deployment")
-}
+// ControllerKind contains the schema.GroupVersionKind for this controller type.
+var ControllerKind = extensions.SchemeGroupVersion.WithKind("Deployment")
```
Unexport this

kargakis commented on pkg/controller/deployment/deployment_controller.go:
```diff
@@ -214,35 +188,48 @@ func (dc *DeploymentController) deleteDeployment(obj interface{}) {
+			return
+		}
+		d, err := dc.dLister.Deployments(rs.Namespace).Get(controllerRef.Name)
+		if err != nil {
```
No need to handle this

kow3ns commented:
Aside from commenting or clarifying some of the parameters to package private methods LGTM

---







*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42175#-:-Ke0UbnM_9sYILPo1Guu:bsi1sbv)*
<!-- Sent from Reviewable.io -->

kow3ns commented:
Reviewed 16 of 16 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/controller/deployment/deployment_controller.go, line 59 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42175#-Ke0P9PAk-T6YLbM04Aj:-Ke0P9PAk-T6YLbM04Ak:b7z22zk) ([raw file](https://github.com/kubernetes/kubernetes/blob/d071b52ac1697728729315b279171dc2a218b0de/pkg/controller/deployment/deployment_controller.go#L59)):*
> ```Go
> 
> // ControllerKind contains the schema.GroupVersionKind for this controller type.
> var ControllerKind = extensions.SchemeGroupVersion.WithKind("Deployment")
> ```

Do we want this to be exported?

---

*[pkg/controller/deployment/sync.go, line 107 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42175#-Ke0S5fSxbcyaY8eDg1_:-Ke0S5fSxbcyaY8eDg1a:b-wdlvt2) ([raw file](https://github.com/kubernetes/kubernetes/blob/d071b52ac1697728729315b279171dc2a218b0de/pkg/controller/deployment/sync.go#L107)):*
> ```Go
> // Note that currently the deployment controller is using caches to avoid querying the server for reads.
> // This may lead to stale reads of replica sets, thus incorrect deployment status.
> func (dc *DeploymentController) getAllReplicaSetsAndSyncRevision(d *extensions.Deployment, rsList []*extensions.ReplicaSet, podMap map[types.UID]*v1.PodList, createIfNotExisted bool) (*extensions.ReplicaSet, []*extensions.ReplicaSet, error) {
> ```

Might want to add comments about these parameters. In particular its not clear immediately that podMap is a map of RS UIDs to their children Pods. It is package private, and the description of the implementation is good. You could change the name of podMap to rsPods or rsPodMap. 

---

*[pkg/controller/deployment/sync.go, line 128 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42175#-Ke0SlvdhUUWjf8oAwem:-Ke0SlvdhUUWjf8oAwen:bbextln) ([raw file](https://github.com/kubernetes/kubernetes/blob/d071b52ac1697728729315b279171dc2a218b0de/pkg/controller/deployment/sync.go#L128)):*
> ```Go
> 
> // rsAndPodsWithHashKeySynced returns the RSes and pods the given deployment targets, with pod-template-hash information synced.
> func (dc *DeploymentController) rsAndPodsWithHashKeySynced(d *extensions.Deployment, rsList []*extensions.ReplicaSet, podMap map[types.UID]*v1.PodList) ([]*extensions.ReplicaSet, *v1.PodList, error) {
> ```

Same comment with respect to podMap, but here the implementation makes it a bit more clear about what is in podMap.

---

*[pkg/controller/deployment/sync.go, line 584 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42175#-Ke0UOBNBRzrt7Astt6u:-Ke0UOBNBRzrt7Astt6v:b3x2k0w) ([raw file](https://github.com/kubernetes/kubernetes/blob/d071b52ac1697728729315b279171dc2a218b0de/pkg/controller/deployment/sync.go#L584)):*
> ```Go
> // isScalingEvent checks whether the provided deployment has been updated with a scaling event
> // by looking at the desired-replicas annotation in the active replica sets of the deployment.
> func (dc *DeploymentController) isScalingEvent(d *extensions.Deployment, rsList []*extensions.ReplicaSet, podMap map[types.UID]*v1.PodList) (bool, error) {
> ```

Same comment about podMap

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42175)*
<!-- Sent from Reviewable.io -->

kargakis commented on pkg/controller/deployment/deployment_controller.go:
```diff
@@ -566,152 +662,29 @@ func (dc *DeploymentController) syncDeployment(key string) error {
 	return fmt.Errorf("unexpected deployment strategy type: %s", d.Spec.Strategy.Type)
 }
-
-// handleOverlap will avoid syncing the newer overlapping ones (only sync the oldest one). New/old is
```
I'm all for removing this but technically since we've started handling overlapping replica sets, it's a disruptive change. I was planning on removing this code in 1.8 when we will stop supporting upgrades from 1.5 - the last release w/o owner references for Deployments.

enisoc replied:
Can you clarify what the disruption is? Are you saying there could be a problem if the cluster is upgraded or downgraded to/from 1.5, if I remove this handleOverlap logic in 1.6?

kargakis replied:
An already overlapping deployment that is ignored by the controller in 1.5 will be able to claim the replica sets of another running deployment after the upgrade.

enisoc replied:
I discussed that upgrade case with @lavalamp for the ControllerRef proposal. He argued that we should not try to prevent this ReplicaSet reassignment upon upgrade to 1.6, because having overlapping Deployments in the first place is unsupported. As a result of that discussion, I added it explicitly in the [Non-Goals](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/controller-ref.md#non-goals) section. It is also called out in the [Upgrading](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/controller-ref.md#upgrading) section, and I will include it in the release note.

kargakis replied:
The reference in the controller ref proposal is definitely not discoverable for admins that want to upgrade their clusters. Release note sgtm

enisoc replied:
I have added `release-note-action-required` to this and the [StatefulSet PR](https://github.com/kubernetes/kubernetes/pull/42080), and specified that the user must ensure overlapping Deployments/StatefulSets don't exist at the time of upgrade.

kargakis replied:
There is also an e2e test that tests overlapping deployments that you'll want to remove, a couple of utilities/annotations in the util directory of the deployment controller, and an annotation check in the kubectl deployment reaper.

kargakis commented on pkg/controller/deployment/deployment_controller.go:
```diff
@@ -446,23 +475,82 @@ func (dc *DeploymentController) handleErr(err error, key interface{}) {
+// getPodCountForReplicaSets scans the list of all Pods and returns the total
+// number of Pods controlled by all ReplicaSets in rsList, based on the Pod's
+// ControllerRef.
+func (dc *DeploymentController) getPodCountForReplicaSets(namespace string, rsList []*extensions.ReplicaSet) (int, error) {
```
Seems like it has a lot of duplicate code. I would be in favor of dropping it and simply estimating the count in the call site (use `getPodMapForReplicaSets`)

enisoc replied:
Done.

kargakis commented on pkg/controller/deployment/deployment_controller.go:
```diff
@@ -446,23 +475,82 @@ func (dc *DeploymentController) handleErr(err error, key interface{}) {
+// RS UID to Pods controlled by that RS, based on the Pod's ControllerRef.
+func (dc *DeploymentController) getPodMapForReplicaSets(namespace string, rsList []*extensions.ReplicaSet) (map[types.UID]*v1.PodList, error) {
+	// List all Pods.
+	pods, err := dc.podLister.Pods(namespace).List(labels.Everything())
```
Why all pods? Just use the deployment selector.

enisoc replied:
My goal was to rely only on ControllerRef to find the Pods owned by the given ReplicaSets, for separation of concerns: the ReplicaSet is in charge of marking what it owns. The Deployment trusts whatever the ReplicaSet puts its mark on.

In particular, if we use the Deployment selector to get the list of Pods, and then filter by ControllerRef, we may leave out a Pod that has the ControllerRef but doesn't match the selector. This means Deployment is making decisions in parallel about which Pods a ReplicaSet owns, rather than trusting the ReplicaSet to communicate ownership via ControllerRef.

kargakis replied:
The pods that match in the controller ref but don't match in labels are invalid and the rs controller will eventually remove the controller ref, right? The deployment controller shouldn't care about that and it should handle only valid pods.

enisoc replied:
@kargakis That makes sense. I also discussed with bgrant and he agreed with you. I will change back to using the Deployment selector, followed by the ControllerRef filtering.

kargakis commented on pkg/controller/deployment/deployment_controller.go:
```diff
@@ -335,8 +357,16 @@ func (dc *DeploymentController) deletePod(obj interface{}) {
-		podList, err := dc.listPods(d)
-		if err == nil && len(podList.Items) == 0 {
+		// Sync if this Deployment now has no more Pods.
+		rsList, err := dc.getReplicaSetsForDeployment(d)
```
Technically and based on the api doc for the deployment selector, we should be able to list all pods managed by a deployment simply by using the deployment selector. Unfortunately, we don't respect this in many places. Mind opening a separate issue to make sure this gets fixed?

enisoc replied:
The guarantee that "we should be able to list all pods managed by a deployment simply by using the deployment selector" seems to conflict with ControllerRef. Or did you mean the weaker guarantee, "all pods managed by a deployment must match the deployment's selector"?

kargakis replied:
I don't think there is a conflict. If the labels don't match, the controller ref is invalid thus removed.

kargakis replied:
And replica set selectors are always a superset of the deployment selector.

enisoc replied:
As discussed in the other thread, I'm now in agreement that Deployment should only care about Pods that match its selector. However, I still think it's important in this function to filter both by Deployment selector AND by ControllerRef matching a ReplicaSet that we own. Otherwise, we will count Pods that we shouldn't be counting, and could get stuck waiting.

Is that what you have in mind?

kargakis replied:
Yes, respecting owner references is the reason for this PR and I am all for that ;)

janetkuo commented:
Assigning @caesarxuchao so that he can also take a look, cc @krmayankk as well

Commit:
Deployment: Always set BlockOwnerDeletion in ControllerRef.

kargakis commented on pkg/controller/deployment/deployment_controller.go:
```diff
@@ -214,50 +188,64 @@ func (dc *DeploymentController) deleteDeployment(obj interface{}) {
+	if rs.DeletionTimestamp != nil {
+		// On a restart of the controller manager, it's possible for an object to
+		// show up in a state that is already pending deletion.
+		dc.deleteReplicaSet(rs)
```
I don't think we care if this is not owned by a Deployment.

enisoc replied:
That check is done inside `deleteReplicaSet()`. Is there a benefit to doing it before calling too?

kargakis replied:
Nvmd, this is fine.

kargakis commented on pkg/controller/deployment/deployment_controller.go:
```diff
@@ -272,16 +260,46 @@ func (dc *DeploymentController) updateReplicaSet(old, cur interface{}) {
-	glog.V(4).Infof("ReplicaSet %s updated.", curRS.Name)
-	if d := dc.getDeploymentForReplicaSet(curRS); d != nil {
+
+	labelChanged := !reflect.DeepEqual(curRS.Labels, oldRS.Labels)
```
Move this next to where it's used.

kargakis commented on pkg/controller/deployment/sync.go:
```diff
@@ -632,8 +592,12 @@ func calculateStatus(allRSs []*extensions.ReplicaSet, newRS *extensions.ReplicaS
+//
+// rsList should come from getReplicaSetsForDeployment(d).
+// podMap should come from getPodMapForReplicaSets(rsList).
+// These are passed around to avoid repeating expensive API calls.
```
This comment is not true, since everything is coming from a cache and not from direct api calls.

enisoc replied:
The potentially expensive part is adoption and orphaning which happens in `getReplicaSetsForDeployment()`, and which I don't want to happen more than once per sync, from a semantic perspective.

The caching of the result from `getPodMapForReplicaSets()` is because it (for now at least; being discussed elsewhere in the PR) lists all pods regardless of selector. If you think this is premature optimization, I can remove `podMap` and just pass the `rsList` around.

kargakis replied:
> The potentially expensive part is adoption and orphaning which happens in getReplicaSetsForDeployment(), and which I don't want to happen more than once per sync, from a semantic perspective.

The adoption/orphaning part was happening separately prior to your PR, exactly for that reason:) We just didn't respect already set owner references.

kargakis replied:
The only mutating action we were doing before is the relabeling of replica sets and pods but that shouldn't happen only in rare cases (adopting old replica sets maybe).

kargakis replied:
The only mutating action we were doing before is the relabeling of replica sets and pods but that should happen only in rare cases (adopting old replica sets maybe).

kargakis replied:
Anyway, I am fine with what you are doing, I have listed it as an item in an old issue of mine (https://github.com/kubernetes/kubernetes/issues/31780).

kargakis commented:
@enisoc @kubernetes/sig-apps-misc this is a fairly involved change for the deployment controller and I am not sure that we really need it for 1.6. Deployment owner references being a special case (they map to replica sets and not pods) and the amount of changes in the controller make me believe we should punt this for 1.7. For all I know we have regressed on much smaller fixes in the past.

kargakis commented:
> @enisoc @kubernetes/sig-apps-misc this is a fairly involved change for the deployment controller and I am not sure that we really need it for 1.6. Deployment owner references being a special case (they map to replica sets and not pods) and the amount of changes in the controller make me believe we should punt this for 1.7. For all I know we have regressed on much smaller fixes in the past.

OTOH, this will need moar reviews and testing if we want it for 1.6

janetkuo commented on pkg/controller/deployment/deployment_controller.go:
```diff
@@ -174,28 +170,6 @@ func (dc *DeploymentController) updateDeployment(old, cur interface{}) {
 	curD := cur.(*extensions.Deployment)
 	glog.V(4).Infof("Updating deployment %s", oldD.Name)
 	dc.enqueueDeployment(curD)
-	// If the selector of the current deployment just changed, we need to requeue any old
```
If we're removing `handleOverlap` we can as well remove the selector update annotation we add when `PrepareForUpdate` in deployment/strategy.go

kargakis replied:
Yes.

janetkuo commented on pkg/controller/deployment/deployment_controller_test.go:
```diff
@@ -544,6 +356,469 @@ func TestPodDeletionDoesntEnqueueRecreateDeployment(t *testing.T) {
+	}
+}
+
+func TestGetReplicaSetsForDeploymentAdopt(t *testing.T) {
```
Can we combine this with the below release test (i.e. add the to-be-released rs to this test)?

janetkuo commented:
`testOverlappingDeployment` in e2e test needs to be changed too

Commit:
Deployment: Filter Pods by Deployment selector in addition to ControllerRef.

Deployment should ignore Pods that don't match the selector, even if
they have a ControllerRef pointing to one of the ReplicaSets it owns.
The ReplicaSet itself will orphan the Pod as soon as it syncs.

Commit:
Deployment: Update overlapping e2e test for ControllerRef.

Commit:
Deployment: Fix data race in unit tests.

kargakis commented:
@mfojtik

Commit:
Deployment: Check that ControllerRef UID matches.

Commit:
Deployment: Consolidate Adopt/Release unit tests.

Commit:
Deployment: Remove Overlap and SelectorUpdate annotations.

These are not used anymore since ControllerRef now protects against
fighting between controllers with overlapping selectors.

enisoc commented:
@k8s-bot unit test this
@k8s-bot kops aws e2e test this
@k8s-bot gce etcd3 e2e test this
@k8s-bot non-cri e2e test this

enisoc commented:
@k8s-bot kops aws e2e test this

kargakis commented on pkg/kubectl/stop.go:
```diff
@@ -441,11 +441,6 @@ func (reaper *DeploymentReaper) Stop(namespace, name string, timeout time.Durati
-	}
-
 	// Stop all replica sets.
 	selector, err := metav1.LabelSelectorAsSelector(deployment.Spec.Selector)
```
I guess we don't want to stop replica sets owned by other deployments.

kargakis commented on pkg/controller/deployment/deployment_controller.go:
```diff
@@ -447,23 +487,56 @@ func (dc *DeploymentController) handleErr(err error, key interface{}) {
+// ControllerRef by adopting and orphaning.
+// It returns the list of ReplicaSets that this Deployment should manage.
+func (dc *DeploymentController) getReplicaSetsForDeployment(d *extensions.Deployment) ([]*extensions.ReplicaSet, error) {
+	rsList, err := dc.rsLister.ReplicaSets(d.Namespace).List(labels.Everything())
```
Add a comment mentioning that we list all ReplicaSets because we also want to orphan.

kargakis replied:
Or just enhance the function comment

Commit:
Deployment: Filter by ControllerRef in Reaper.

We don't want to delete ReplicaSets we don't own.

janetkuo reviewed (approved):
This change LGTM.

kargakis commented:
@k8s-bot cvm gce e2e test this
@k8s-bot non-cri e2e test this
@k8s-bot non-cri node e2e test this

kargakis commented:
/approve
/lgtm

enisoc commented:
@k8s-bot cvm gce e2e test this
@k8s-bot non-cri e2e test this

janetkuo commented:
/approve

caesarxuchao commented:
/approve

Approve the changes to client.

smarterclayton commented:
/approve

kargakis commented on pkg/controller/deployment/util/deployment_util.go:
```diff
@@ -575,28 +566,94 @@ type rsListFunc func(string, metav1.ListOptions) ([]*extensions.ReplicaSet, erro
+	}
+	// Only include those whose ControllerRef matches the Deployment.
+	owned := make([]*extensions.ReplicaSet, 0, len(all))
+	for _, rs := range all {
```
@enisoc @janetkuo this is going to return empty results for some Deployments after a cluster upgrade (the Deployment controller has yet to adopt all RSs) so it will end up creating duplicate ReplicaSets [again](https://github.com/kubernetes/kubernetes/issues/40415), right? :/ We should probably relax those checks for 1.6. It would also fix the upgrade test issue in https://github.com/kubernetes/kubernetes/issues/42738

kargakis replied:
Hm, seems like the controller is making sure to adopt all replica sets first before proceeding to handle them which is good.

enisoc replied:
Right, the controller itself now uses a separate code path (ControllerRefManager). This one is only used from outside the controller.
